pyrelational/pipeline/generic_pipeline.py

```diff
@@ -0,0 +1,321 @@
+"""This module defines the acquisition manager which utilises
+the data manager, sampling functions, and model to create acquisition
+functions and general arbiters of the active learning pipeline
+"""
+import logging
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from typing import Dict, List, Optional
+
+import numpy as np
+import pandas as pd
+from tabulate import tabulate
+from torch.utils.data import DataLoader
+
+from pyrelational.data.data_manager import GenericDataManager
+from pyrelational.models.generic_model import GenericModel
+from pyrelational.oracle.generic_oracle import GenericOracle
+from pyrelational.strategies.generic_al_strategy import GenericActiveLearningStrategy
+
+logger = logging.getLogger()
+
+
+class GenericPipeline(ABC):
+    """The pipeline facillitates the communication between
+    - DataManager
+    - Model,
+    - ALStrategy,
+    - Oracle (Optional)
+
+    To enact a generic ative learning cycle.
+
+    :param data_manager: an pyrelational data manager
+            which keeps track of what has been labelled and creates data loaders for
+            active learning
+    :param model: A pyrelational model
+            which serves as the machine learning model for the data in the
+            data manager
+    :param strategy: A pyrelational active learning strategy
+            implements the informativeness measure and the selection algorithm being used
+    :param oracle: An oracle instance
+            interfaces with various concrete oracles to obtain labels for observations
+            suggested by the strategy
+    """
+
+    def __init__(
+        self,
+        data_manager: GenericDataManager,
+        model: GenericModel,
+        strategy: GenericActiveLearningStrategy,
+        oracle: Optional[GenericOracle] = None,
+    ):
+        super(GenericPipeline, self).__init__()
+        self.data_manager = data_manager
+        self.model = model
+        self.strategy = strategy
+        self.oracle = oracle
+
+        # Mediator pattern for strategy
+        self.strategy._pipeline = self
```

Comment:
not sure this is actually used in the strategies so far as well

Reply:
Apologies I initially tried building it using a mediator design pattern, but went in favour of using the current  _filter_kwargs pattern, so this is a deprecated artifact. This will be removed in the next commit.